android/src/main/java/com/guilded/gg/AudioInputController.java

```diff
@@ -0,0 +1,132 @@
+
+package com.guilded.gg;
+
+import android.media.AudioFormat;
+import android.media.AudioRecord;
+import android.media.MediaRecorder;
+import android.util.Log;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AudioInputController {
+    public interface AudioInputControllerListener {
+        void onProcessSampleData(ByteBuffer buffer);
+        void onProcessingError(String error);
+    }
+
+    public static final int AudioChannelConfig = AudioFormat.CHANNEL_IN_MONO;
+    public static final int AudioSampleFormat = AudioFormat.ENCODING_PCM_8BIT;
+
+
+    private AudioInputControllerListener listener;
+
+    private AudioRecord recorder = null;
+    private Thread recordingThread = null;
+    private final AtomicBoolean recordingInProgress = new AtomicBoolean(false);
+
+    private int sampleRate;
+    private int bufferSize;
+
+    private AudioInputController() {
+    }
+
+    private static class AudioInputControllerSingleton {
+        private static final AudioInputController sharedInstance = new AudioInputController();
+    }
+
+    public static AudioInputController getInstance() {
+        return AudioInputControllerSingleton.sharedInstance;
+    }
+
+    public int sampleRate() {
+        return this.sampleRate;
+    }
+
+    public void start() {
+        recorder = new AudioRecord(MediaRecorder.AudioSource.DEFAULT, sampleRate,
+                AudioInputController.AudioChannelConfig, AudioInputController.AudioSampleFormat, bufferSize);
+
+        recorder.startRecording();
+        recordingInProgress.set(true);
+
+        recordingThread = new Thread(new RecordingRunnable(this), "Voice detector thread");
+        recordingThread.start();
+
+    }
+
+
+    public void stop() {
+        if (recorder != null) {
+            recordingInProgress.set(false);
+            recorder.release();
+            recorder = null;
+            recordingThread = null;
+        }
+    }
+
+    // Assign the listener implementing events interface that will receive the events
+    public void setAudioInputControllerListener(AudioInputControllerListener listener) {
+        this.listener = listener;
+    }
+
+    public void prepareWithSampleRate(int desiredSampleRate) {
+        int sampleRate = 44100; // default sample rate: supported on most android devices
```

Comment:
what happens on android devices where this is not supported? does voice just not work? if so, should we handle those cases with some sort of compatibility banner or something?

Reply:
Voice will work find, but VAD may not work as expected. I don't have an array of HW devices to try this on, but from reading, we get like 85%+